source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,166 @@ namespace aspect
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell != endc; ++cell)
+      {
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (local_dof_indices);
+            fe_values_12.reinit (cell);
+            fe_values_21.reinit (cell);
+
+            fe_values_12[field].get_function_values(solution, values_12);
+            fe_values_21[field].get_function_values(solution, values_21);
+
+            //Find the local max and local min
+            double min_advection_local =  std::numeric_limits<double>::max();
+            double max_advection_local = -std::numeric_limits<double>::max();
+            for (unsigned int q = 0; q < n_q_points_12; ++q)
+              {
+                min_advection_local = std::min<double> (min_advection_local, values_12[q]);
+                max_advection_local = std::max<double> (max_advection_local, values_12[q]);
+              }
+            for (unsigned int q = 0; q < n_q_points_21; ++q)
+              {
+                min_advection_local = std::min<double> (min_advection_local, values_21[q]);
+                max_advection_local = std::max<double> (max_advection_local, values_21[q]);
+              }
+
+            //Find the trouble cell
+            if (min_advection_local < min_advection_exact_global
+                || max_advection_local > max_advection_exact_global)
+              {
+                //Compute the cell area and cell advection average
+                double local_area = 0;
+                double local_advection_average = 0;
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_area += fe_values_12.JxW(q);
+                    local_advection_average += values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_advection_average /= local_area;
+                /*
+                 * Define theta: a scaling constant used to correct the old solution by the formula
+                 *   new_value = theta * (old_value-old_solution_cell_average)+old_solution_cell_average
+                 * where theta \in [0,1] defined as below.
+                 * After the correction, the new solution does not exceed the user-given
+                 * exact global maximum/minimum values. Meanwhile, the new solution's cell average
+                 * equals to the old solution's cell average.
+                 */
+                double theta = std::min<double>
+                               (1, abs((max_advection_exact_global-local_advection_average)
+                                       /(max_advection_local-local_advection_average)));
+                theta = std::min<double>
+                        (theta, abs((min_advection_exact_global-local_advection_average)
+                                    /(min_advection_local-local_advection_average)));
+                //Modify the numerical solution at freedoms of advection field
```

Comment:
maybe rephrase this as "modify the advection degrees of freedom of the numerical solution